### PR TITLE
Nbody demo improvements

### LINKED ISSUE #40: Use OpenGL for nbody demo
#36 added an nbody demo. It would be nice to glium instead of SDL though.

### PR DESCRIPTION
**Pull request by bjz**
Uses `cgmath` for maths, `glium` for graphics, and `docopt` for arg parsing.

Fixes #40.

### CODE REVIEW

File: demo/nbody/src/visualize.rs [hunk 1]
```diff
@@ -1,62 +1,180 @@
...
+        3,  7, 11,
+        3,  9,  6,
+        4,  8,  9,
+        5, 11, 10,
+        6,  9,  8,
+        7, 10, 11,
+    ];
+
+    (vertices, indices)
+}
```
Comment (bjz) [inline demo/nbody/src/visualize.rs]: Couldn't be bothered making a sphere function - had this code lying around....

File: demo/nbody/src/visualize.rs [hunk 2]
```diff
@@ -1,62 +1,180 @@
...
+        };
+
+        let mut target = display.draw();
+
+        let (width, height) = target.get_dimensions();
+        let aspect = width as f32 / height as f32;
+
+        let proj = cgmath::perspective(Rad::full_turn() / 6.0, aspect, 0.1, 3000.0);
+        let view = Matrix4::look_at(Point3::new(10.0, 10.0, 10.0), Point3::origin(), Vector3::unit_z());
+        let view_proj: [[f32; 4]; 4] = (proj * view).into();
```
Comment (bjz) [inline demo/nbody/src/visualize.rs]: The camera could be tweaked more, but you can see most of it.

File: demo/nbody/src/main.rs [hunk 3]
```diff
@@ -1,111 +1,123 @@
...
-
-    let par_time = {
+fn run_benchmarks(run_par: bool, run_seq: bool, bodies: usize, ticks: usize) {
+    let par_time = if run_par {
         let mut rng = XorShiftRng::from_seed([0, 1, 2, 3]);
-        let mut benchmark = NBodyBenchmark::new(num_bodies, &mut rng);
+        let mut benchmark = NBodyBenchmark::new(bodies, &mut rng);
         let par_start = time::precise_time_ns();
+
+        let mut progress = ProgressBar::new(ticks);
```
Comment (lilianmoraru) [inline demo/nbody/src/main.rs]: Isn't this also benchmarking the ProgressBar?
Output takes a lot of time(Note that I don't exactly know how `pbr` works but my guess is that when you call `progress.inc()` it blocks to write to the output).

File: demo/nbody/src/main.rs [hunk 4]
```diff
@@ -1,111 +1,123 @@
...
+
+        try!(writeln!(f, "  --bodies {}", self.flag_bodies));
+        try!(writeln!(f, "  --ticks {}", self.flag_ticks));
+
+        Ok(())
+    }
+}
 
 fn main() {
-    let mut num_bodies = DEFAULT_NUM_BODIES;
```
Comment (lilianmoraru) [inline demo/nbody/src/main.rs]: I like how `docopt` makes all of this look so much better :)

**Comment by lilianmoraru** [comment]
I haven't read the book for a long time(I think it was mentioned there) that I already forgot you can do things like `'main: loop { ... code ...; break 'main; }` :)

File: demo/nbody/src/main.rs (hunk 3 above)
Comment (bjz) [inline demo/nbody/src/main.rs]: Hmm, you might be right. Although `ticks` is quite small. I can remove it though, just felt nicer to have something to look at :)

**Comment by bjz** [comment]
Yeah, I only remember it because it is useful for breaking out of nested game/input event loops. :)

**Comment by nikomatsakis** [comment]
Hmm, building locally I get errors like this:

```
/Users/nmatsakis/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/glutin-0.4.7/src/api/cocoa/mod.rs:148:18: 148:62 error: the trait `objc::encode::Encode` is not implemented f\
or the type `*mut libc::c_void` [E0277]
/Users/nmatsakis/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/glutin-0.4.7/src/api/cocoa/mod.rs:148             decl.add_ivar::<*mut libc::c_void>("glutinState");
                                                                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

any idea what is going on?

**Comment by nikomatsakis** [comment]
cargo update fixed it.

**Comment by nikomatsakis** [comment]
@bjz ok, to get this to work, I had to `s/cmd_benchmark/cmd_bench`, but once I did that, it looks pretty awesome. I've not looked at the actual diff yet -- but I have one more request. Can you make it so that the visualizer reports FPS, and can be changed from par to seq mode? I mean, I'll merge it either way, but that'd be the icing on the cake. ;)

File: demo/nbody/src/main.rs (hunk 3 above)
Comment (nikomatsakis) [inline demo/nbody/src/main.rs]: Yeah, I'm not sure about benchmarking the progress bar.

File: demo/nbody/src/main.rs [hunk 5]
```diff
@@ -1,111 +1,123 @@
+ #![feature(augmented_assignments)]
```
Comment (nikomatsakis) [inline demo/nbody/src/main.rs]: Hmm, I'm a bit disappointed to take this off of stable. But it's just a demo I guess.

File: demo/nbody/src/nbody.rs [hunk 6]
```diff
@@ -230,22 +223,22 @@ fn next_velocity(time: usize, prev: &Body, bodies: &[Body]) -> (Vector, Vector)
                 // normalize the distance vector
                 let d = (r / length) * c;
 
-                acc.increment(d);
-                acc2.decrement(d);
+                acc += d;
+                acc2 -= d;
```
Comment (nikomatsakis) [inline demo/nbody/src/nbody.rs]: And this IS so much prettier. :)

File: demo/nbody/src/main.rs (hunk 3 above)
Comment (nikomatsakis) [inline demo/nbody/src/main.rs]: I definitely saw a drop in parallel speedup locally when using the progress bar. Makes sense. Amdahl's law and all that. (3.9x to 3.5x on a 4core)

File: demo/nbody/src/main.rs (hunk 5 above)
Comment (bjz) [inline demo/nbody/src/main.rs]: Only thing is the `*=` operators from cgmath. I could just use regular re-assignments, which I'm guessing should optimize ok, which mean we could drop cgmath's `"unstable"` feature flag.

**Comment by bjz** [comment]
FPS counter would take a bit more work (me figuring out fonts and stuff).

File: demo/nbody/src/nbody.rs (hunk 6 above)
Comment (bjz) [inline demo/nbody/src/nbody.rs]: At the expense of `#![feature(augmented_assignments)]` :(

**Comment by nikomatsakis** [comment]
@bjz can you make it switch from par to seq by typing `p` or `s` ?

**Comment by nikomatsakis** [comment]
@bjz yeah let's keep the augmented-assignments, this is just a demo, and they are on their way to being stable anyway.

**Comment by nikomatsakis** [comment]
Ah, good enough. Thanks @bjz, this is great. :)

**Pull request merged**